From a07cdc9c2e1dc0a2a6a9c44b3cb6e2defb7bff63 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Thu, 09 Jan 2020 15:54:45 +0200
Subject: [PATCH] [SW-6739] arcboot: fix memcpy alignment exception

in synopsis secureboot library after decrypting the
signature the library would try to extract the hash
to verify. the code in would memcpy from unaligned
memory address causing exception on the arc.
work around issue for now by disabling exceptions
from unaligned memory access.

Change-Id: If56ba35e6ed676c6f200671b83f81b2aa0706dd5
---

diff --git a/common/crt0_rom.S b/common/crt0_rom.S
index e247fcd..1a43094 100644
--- a/common/crt0_rom.S
+++ b/common/crt0_rom.S
@@ -12,6 +12,14 @@
 ;; *******************************************************************************************************************
 .file "crt0_rom.S"
 
+.macro CPU_EARLY_SETUP
+
+	; Disable alignment checking
+	lr	r5, [status32]
+	bset	r5, r5, 19
+	kflag	r5
+.endm
+
 .extern main
 
 .section .ivt, "a", @progbits
@@ -64,6 +72,7 @@
 ;   instruction
 ; - loop body should have at least two instruction words
 __start:
+	CPU_EARLY_SETUP
 	mov gp, @__SDATA_BEGIN__
 
 	mov_s r2, @__sbss_start	; r2 = start of the bss section
